src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -62,4 +65,91 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
         instance.AddRange(items);
         return disposer;
     }
+
+    /// <summary>
+    /// Main utility for searching across items in a solution.  The actual code to search the item should be provided in
+    /// <paramref name="callback"/>.  Each item in <paramref name="items"/> will be processed using
+    /// <code>Parallel.ForEachAsync</code>, allowing for parallel processing of the items, with a preference towards
+    /// earlier items.
+    /// </summary>
+    private static async Task PerformParallelSearchAsync<T>(
+        IEnumerable<T> items,
+        Func<T, Action<RoslynNavigateToItem>, CancellationToken, ValueTask> callback,
+        Func<ImmutableArray<RoslynNavigateToItem>, Task> onItemsFound,
+        CancellationToken cancellationToken)
```

Comment:
this is hte core utility that all 3 nav-to search functions end up deferring to (search normal docs.  search generated docs.  search cached docs).